### PR TITLE
FC displacement

### PR DESCRIPTION
**Pull request by paolafer**
This PR displaces the field cage volumes in `xy`, trying to reproduce the real displacement at LSC as much as possible. This displacement is caused by having two HDPE layers only in one sector of the cylinder, and not all over the place.
Modifications have been made in order to have the `ACTIVE` volume centred in the TPC axis.

### CODE REVIEW

File: source/geometries/GeometryBase.h [hunk 1]
```diff
@@ -40,15 +40,18 @@ namespace nexus {
     /// Returns the 3 dimensions of the geometry (x, y, z)
     G4ThreeVector GetDimensions();
 
-    /// Returns true if geometry has a drift field
-    G4bool GetDrift() const;
```
Comment (MiryamMV) [inline source/geometries/GeometryBase.h:50]: Is this related to the field cage displacement?

File: source/geometries/Next100.cc [hunk 2]
```diff
@@ -43,6 +43,8 @@ namespace nexus {
     gate_tracking_plane_distance_(25. * mm + grid_thickness_), // Jordi = 1.5 (distance TP plate-anode ring) + 13.5 (anode ring thickness) + 10 (EL gap)
     gate_sapphire_wdw_distance_  (1458.8 * mm - grid_thickness_), // Jordi
     ics_ep_lip_width_ (55. * mm), // length of the step cut out in the ICS, in the EP side
+    fc_displ_x_ (-3.7 * mm), // displacement of the field cage volumes from 0
+    fc_displ_y_ (-6.4 * mm), // displacement of the field cage volumes from 0
```
Comment (MiryamMV) [inline source/geometries/Next100.cc:47]: How have these displacements been determined? We're they measured in the lab?

File: source/geometries/Next100.cc [hunk 3]
```diff
@@ -130,49 +136,49 @@ namespace nexus {
...
       new G4PVPlacement(0, castle_pos, shielding_logic,
                         "LEAD_BOX", hallA_logic_, false, 0);
-      new G4PVPlacement(0, gate_pos - castle_pos, hallA_logic_,
+      new G4PVPlacement(0, -coord_origin_ - castle_pos, hallA_logic_,
                         "Hall_A", lab_logic_, false, 0, false);
     }
     else {
-      new G4PVPlacement(0, gate_pos, shielding_logic, "LEAD_BOX", lab_logic_,
-                        false, 0);
+      new G4PVPlacement(0, -coord_origin_, shielding_logic,
```
Comment (MiryamMV) [inline source/geometries/Next100.cc:183]: This is me not understanding the the geometry, why do we add a minus sign to `coord_origin_`?

File: source/geometries/Next100.cc [hunk 4]
```diff
@@ -265,8 +271,7 @@ namespace nexus {
 		  "Unknown vertex generation region!");
     }
 
-    G4ThreeVector displacement = G4ThreeVector(0., 0., -gate_zpos_in_vessel_);
-    vertex = vertex + displacement;
+    vertex = vertex - coord_origin_;
```
Comment (MiryamMV) [inline source/geometries/Next100.cc:294]: Same here. Why a minus and not a plus?

File: source/geometries/Next100FieldCage.cc [hunk 5]
```diff
@@ -72,8 +72,8 @@ Next100FieldCage::Next100FieldCage(G4double grid_thickn):
   // cathode_sapphire_wdw_dist_(257.8 * mm),
 
   // external to teflon (hdpe + rings + holders)
-  hdpe_tube_int_diam_ (1093. * mm),
-  hdpe_tube_ext_diam_ (1105.4 * mm),
+  hdpe_tube_int_diam_ (1093. * mm - 11. * mm), // we keep the thickness, but we make it smaller to be closer to the staves
+  hdpe_tube_ext_diam_ (1105.4 * mm - 11. * mm),
```
Comment (MiryamMV) [inline source/geometries/Next100FieldCage.cc]: Is this 11 derived from the values of the displacement?

File: source/geometries/Next100Vessel.h [hunk 6]
```diff
@@ -96,6 +101,14 @@ namespace nexus {
     gate_tp_distance_ = distance;
   }
 
+  inline void Next100Vessel::SetGateZpos(G4double pos){
+    gate_z_pos_ = pos;
+  }
+
+  inline G4double Next100Vessel::GetGateZpos(){
```
Comment (MiryamMV) [inline source/geometries/Next100Vessel.h:108]: This for sure has a reason but I can't get my head around it. Why do we define this here and not in the Next100FieldCage class, for example.

File: source/geometries/Next100EnergyPlane.cc [hunk 7]
```diff
@@ -207,7 +207,8 @@ namespace nexus {
     G4double stand_out_length =
       sapphire_window_thickn_ + tpb_thickn_ + optical_pad_thickn_ + pmt_stand_out_;
 
-    copper_plate_posz_ = GetELzCoord() + gate_sapphire_wdw_dist_ + stand_out_length + copper_plate_thickn_/2.;
+    copper_plate_posz_ = GetCoordOrigin()[2]
```
Comment (MiryamMV) [inline source/geometries/Next100EnergyPlane.cc]: In other clases (like Next100FieldCage.cc L216-224) you use `.z()` to extract the z component of the G4ThreeVector. Is it possible to do the same here?

File: source/geometries/Next100FieldCage.cc [hunk 8]
```diff
@@ -330,9 +333,16 @@ void Next100FieldCage::BuildActive()
   drift_region->AddRootLogicalVolume(active_logic);
 
   /// Vertex generator
-  active_gen_ = new CylinderPointSampler2020(0., active_ext_radius_, active_length_/2.,
+  active_gen_radius_ =
+    std::max(gate_int_diam_/2. - GetCoordOrigin().x(),
+             gate_int_diam_/2. - GetCoordOrigin().y());
+
+  active_gen_radius_ = std::max(active_gen_radius_, active_ext_radius_);
```
Comment (MiryamMV) [inline source/geometries/Next100FieldCage.cc:340]: I don't understand. Won't the external radius always be greater than the internal radius?

File: source/geometries/Next100FieldCage.cc [hunk 9]
```diff
@@ -516,7 +534,7 @@ void Next100FieldCage::BuildELRegion()
   if (elfield_) {
     /// Define EL electric field
     UniformElectricDriftField* el_field = new UniformElectricDriftField();
-    G4double global_el_gap_zpos = el_gap_zpos_ - GetELzCoord();
+    G4double global_el_gap_zpos = el_gap_zpos_ - GetCoordOrigin()[2];
```
Comment (MiryamMV) [inline source/geometries/Next100FieldCage.cc]: Same as before. I personally find it more intuitive.

File: source/geometries/Next100FieldCage.cc [hunk 10]
```diff
@@ -757,17 +786,20 @@ void Next100FieldCage::BuildFieldCage()
...
   }
 
   // Ring vertex generator
   G4double ring_gen_lenght = first_ring_buff_z_pos + ring_thickn_/2. - (posz - ring_thickn_/2.);
   G4double ring_gen_zpos = first_ring_buff_z_pos + ring_thickn_/2. - ring_gen_lenght/2.;
-  ring_gen_ = new CylinderPointSampler2020(ring_int_diam_/2., ring_ext_diam_/2., ring_gen_lenght/2.,
-                                           0., twopi, nullptr,
-                                           G4ThreeVector(0., 0., ring_gen_zpos));
+  ring_gen_ =
+    new CylinderPointSampler2020(ring_int_diam_/2., ring_ext_diam_/2., ring_gen_lenght/2.,
```
Comment (MiryamMV) [inline source/geometries/Next100FieldCage.cc]: This is not related to the PR but I just realized that `ring_gen_lenght` has a typo. And it might be spreaded throughout the whole code.

File: source/geometries/GeometryBase.h (hunk 1 above)
Comment (paolafer) [reply source/geometries/GeometryBase.h:50]: This was an obsolete function, which was not used practically in the code. Since this PR adds a displacement in the 3 dimensions (instead of only in `z`), I took the chance of restructuring the code using a generic 3D displacement. This was something I had already started in a previous PR, which I have closed now, since it is being implemented here.

File: source/geometries/Next100.cc (hunk 2 above)
Comment (paolafer) [reply source/geometries/Next100.cc:47]: I have used the measurements of the distances between the ICS bars and the teflon panels as a reference. These distances can be found in the picture [here](https://github.com/next-exp/nexus/wiki/The-NEXT-100-geometry). The `fc_displ_x_` and `fc_displ_y_` variables have been adjusted to get as closest to the real measurements as possible, without producing overlaps in the geometry. The result is not perfect, but close enough.

File: source/geometries/Next100.cc (hunk 3 above)
Comment (paolafer) [reply source/geometries/Next100.cc:183]: Because if the field cage is displaced by a quantity from the origin, you want to place the whole volume that contains everything in `-` that quantity, to have the field cage in `(0, 0, 0)` as a result.

File: source/geometries/Next100.cc (hunk 4 above)
Comment (paolafer) [reply source/geometries/Next100.cc:294]: Same as above. In the code there was a `+` before simply because the `displacement` variable was already defined with a minus sign.

File: source/geometries/Next100FieldCage.cc (hunk 5 above)
Comment (paolafer) [reply source/geometries/Next100FieldCage.cc]: No, `11 mm` is the distance we need to reduce the diameter of, to get the hdpe tube basically touching the teflon panels. In the picture taken in the lab, the two objects look like touching each other, while in the technical drawings it is not the case. I wanted to keep track of the measurements of the technical drawings, therefore I did keep the numbers in the code and simply subtract a fixed, _ad hoc_, quantity.

File: source/geometries/Next100Vessel.h (hunk 6 above)
Comment (paolafer) [reply source/geometries/Next100Vessel.h:108]: Because the position of the gate with respect to the center of the gas (which is used everywhere else in the code) is calculated in the `Next100Vessel` class, therefore we need a function to get it from that class only.

File: source/geometries/Next100EnergyPlane.cc (hunk 7 above)
Comment (paolafer) [reply source/geometries/Next100EnergyPlane.cc]: Yes, I must have missed it. Done!

File: source/geometries/Next100FieldCage.cc (hunk 9 above)
Comment (paolafer) [reply source/geometries/Next100FieldCage.cc]: Done.

File: source/geometries/Next100FieldCage.cc (hunk 10 above)
Comment (paolafer) [reply source/geometries/Next100FieldCage.cc]: Fixed!

File: source/geometries/Next100FieldCage.cc (hunk 8 above)
Comment (paolafer) [reply source/geometries/Next100FieldCage.cc:340]: The thing is that the active volume is made of the union of two pieces, a polyhedron and a cylinder. On top of that, these pieces are not glued centered on the same axis, but displaced due to the field cage displacement. Therefore, depending on the value of the displacement, it can be that one needs a bigger radius for the generator, than the external radius of the polyhedron. Notice that the generator is a cylinder that needs to overcover all the active volume, to keep the code simple.

File: source/geometries/Next100.cc (hunk 3 above)
Comment (MiryamMV) [reply source/geometries/Next100.cc:183]: I understand now.

File: source/geometries/Next100FieldCage.cc (hunk 5 above)
Comment (MiryamMV) [reply source/geometries/Next100FieldCage.cc]: I don't get it. The teflon panels shouldn't touch the HDPE tube because they have the holder/staves in between.

File: source/geometries/Next100OpticalGeometry.cc [hunk 11]
```diff
@@ -34,11 +34,13 @@ namespace nexus {
     // common variables used in geometry components
     grid_thickness_ (0.1 * mm),
     gate_tracking_plane_distance_(25. * mm + grid_thickness_),
-    gate_sapphire_wdw_distance_  ((1458.8 - 1.3) * mm - grid_thickness_),
+    gate_sapphire_wdw_distance_  (1458.8 * mm - grid_thickness_),
```
Comment (MiryamMV) [inline source/geometries/Next100OpticalGeometry.cc:37]: Why were the 1.3mm removed and what was the purpose of it initially?

File: source/geometries/Next100OpticalGeometry.cc [hunk 12]
```diff
@@ -126,19 +128,19 @@ namespace nexus {
                 "Unknown kind of gas, valid options are: naturalXe, enrichedXe, depletedXe.");
   }
 
-  G4double gas_size = lab_size - 1.*cm;
+  G4double gas_size = lab_size - 10.*cm;
```
Comment (MiryamMV) [inline source/geometries/Next100OpticalGeometry.cc:131]: Was this increased to avoid possible overlaps?

File: source/geometries/Next100TrackingPlane.cc [hunk 13]
```diff
@@ -84,7 +84,7 @@ void Next100TrackingPlane::Construct()
   G4LogicalVolume* copper_plate_logic =
     new G4LogicalVolume(copper_plate_solid, copper, copper_plate_name);
 
-  G4double copper_plate_zpos = GetELzCoord() - gate_tp_dist_ - copper_plate_thickness_/2.;
+  G4double copper_plate_zpos = GetCoordOrigin()[2] - gate_tp_dist_ - copper_plate_thickness_/2.;
```
Comment (MiryamMV) [inline source/geometries/Next100TrackingPlane.cc]: Same as in previous comments.

File: source/geometries/Next100TrackingPlane.cc [hunk 14]
```diff
@@ -100,7 +100,7 @@ void Next100TrackingPlane::Construct()
   sipm_board_geom_->Construct();
   G4LogicalVolume* sipm_board_logic = sipm_board_geom_->GetLogicalVolume();
 
-  G4double zpos = GetELzCoord() - gate_tp_dist_ + sipm_board_geom_->GetThickness()/2.;
+  G4double zpos = GetCoordOrigin()[2] - gate_tp_dist_ + sipm_board_geom_->GetThickness()/2.;
```
Comment (MiryamMV) [inline source/geometries/Next100TrackingPlane.cc]: And same here.

File: source/geometries/Next100FieldCage.cc (hunk 5 above)
Comment (paolafer) [reply source/geometries/Next100FieldCage.cc]: Sorry, I meant the staves, not the teflon panels.

File: source/geometries/Next100FieldCage.cc (hunk 10 above)
Comment (MiryamMV) [reply source/geometries/Next100FieldCage.cc]: Thanks!

File: source/geometries/Next100FieldCage.cc (hunk 5 above)
Comment (MiryamMV) [reply source/geometries/Next100FieldCage.cc]: In this STEP file it seems like they're touching: 
[https://viewer.autodesk.com/id/dXJuOmFkc2sub2JqZWN0czpvcy5vYmplY3Q6YTM2MHZpZXdlci1wcm90ZWN0ZWQvdDE3MDE4NjM3NjhfNTE1MDE0ZGMtMzA1OC00ZWUwLWFjNTQtZDZmZTc2NDllODQ0LnN0ZXA?sheetId=ZTgzMzNkMjUtZDhhOC00NmFjLThlMTYtMWYzODdkMGU5NDZi]

File: source/geometries/Next100OpticalGeometry.cc (hunk 11 above)
Comment (paolafer) [reply source/geometries/Next100OpticalGeometry.cc:37]: It was just a fine tuning based on the technical drawings. After the whole review, it turned out that the correct distance was 1458.8 mm.

File: source/geometries/Next100OpticalGeometry.cc (hunk 12 above)
Comment (paolafer) [reply source/geometries/Next100OpticalGeometry.cc:131]: Yes, after the displacement of the FC, `1 cm` was too small, and the displacement produced overlaps.

File: source/geometries/Next100FieldCage.cc (hunk 5 above)
Comment (paolafer) [reply source/geometries/Next100FieldCage.cc]: There is an incongruence somewhere, because the 2D drawings say that the internal diameter is 1093. mm. I'll have a look!

File: source/geometries/Next100TrackingPlane.cc (hunk 13 above)
Comment (paolafer) [reply source/geometries/Next100TrackingPlane.cc]: Done.

File: source/geometries/Next100TrackingPlane.cc (hunk 14 above)
Comment (paolafer) [reply source/geometries/Next100TrackingPlane.cc]: Done.

File: source/geometries/Next100FieldCage.cc (hunk 5 above)
Comment (paolafer) [reply source/geometries/Next100FieldCage.cc]: Here you can find the technical drawing of hdpe, slide 4: 
[NEXT-100_111522_HDPE_and_staves_PLANOS FABRICACIÓN.pdf](https://github.com/next-exp/nexus/files/13635208/NEXT-100_111522_HDPE_and_staves_PLANOS.FABRICACION.pdf)

As you can see, the internal diameter is 1093 mm. However, with that diameter hdpe is not touching the staves, provided that the active diameter is 984 mm (which has not been measured directly and I don't know what the techincal drawings say).

Comment (paolafer) [reply source/geometries/Next100FieldCage.cc]: If we leave the hdpe diameter as it is in the drawings, the distance would be like in the picture:
Screenshot 2023-12-11 at 14 49 05

Take into account that hdpe is a cylinder, although in the picture it is shown as a polyhedron.

Comment (paolafer) [reply source/geometries/Next100FieldCage.cc]: Shall we keep it like this?

Comment (paolafer) [reply source/geometries/Next100FieldCage.cc]: I realized that the dimensions of the hdpe were those of the original external wrap. I've changed them to those of the internal one, since they are almost touching the staves. I've also updated the distance between two opposite teflon panels, according to the drawings. We know from in-situ measurements that the field cage is not a perfect cylinder, but it is slightly deformed. However, it is impossible to simulate that correctly, therefore we decided to keep the simulation simple.

File: source/geometries/Next100FieldCage.cc [hunk 15]
```diff
@@ -72,8 +72,8 @@ Next100FieldCage::Next100FieldCage(G4double grid_thickn):
   // cathode_sapphire_wdw_dist_(257.8 * mm),
 
   // external to teflon (hdpe + rings + holders)
-  hdpe_tube_int_diam_ (1093. * mm - 11. * mm), // we keep the thickness, but we make it smaller to be closer to the staves
-  hdpe_tube_ext_diam_ (1105.4 * mm - 11. * mm),
+  hdpe_tube_int_diam_ (1081. * mm),
```
Comment (MiryamMV) [inline source/geometries/Next100FieldCage.cc:75]: Maybe it's irrelevant, but the thickness before  and after is not exactly the same.

Comment (paolafer) [reply source/geometries/Next100FieldCage.cc:75]: No, it's not the same. It turns out that the external hdpe cylinder is 6 mm-thick, while the internal one is 6.5 mm-thick, according to the drawing attached above.
Since my understanding is that the internal one has been kept, I think that now it is done correctly.

Comment (MiryamMV) [reply source/geometries/Next100FieldCage.cc:75]: Great!

**Comment by paolafer** [comment]
Hi @MiryamMV, @jmalbos, what's the status of your review?

**Review by MiryamMV — APPROVED** [review]
This PR implements the changes necessary to simulate the field cage displacement successfully. I approve it. Good job @paolafer !
